feat(aws-cdk): allow changing endpoints of the AWS sdk

MathiasKoch opened the pull request:
Allow changing endpoints of the AWS sdk, to support deployment to local environments. 

This feature adds the ability to provide an optional `endpoints` argument to the aws-cdk cli tool.

eg.
```
cdk deploy --endpoints 'cloudformation=http://localhost:4581,s3=http://localhost:4572'
cdk deploy --endpoints '{"cloudformation":"http://localhost:4581","s3":"http://localhost:4572"}'
cdk deploy --endpoints $(cat endpoints.json)
```

The format of the endpoints.json file should be:
```
{
    "cloudformation": "http://localhost:4581",
    "s3": "http://localhost:4572"
}
```

This will allow the CDK cli to deploy and interact with local environments like https://github.com/localstack/localstack 

closes #1989

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

<!-- 
Please read the contribution guidelines and follow the pull-request checklist:
https://github.com/aws/aws-cdk/blob/master/CONTRIBUTING.md
 -->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* Result: SUCCEEDED
* [Build Logs](https://dfq49x5zyi.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=dd4a73b3-70ea-46f4-be29-d65306978fea%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

nmussy commented:
It might be nice to support a shorthand syntax, similar to the AWS CLI:

```sh
cdk deploy --endpoints 'cloudformation=http://localhost:4581,s3=http://localhost:4572'
cdk deploy --endpoints '{"cloudformation":"http://localhost:4581","s3":"http://localhost:4572"}'
```

As for the file, I don't think it's necessary for the CDK to handle reading them. You could just as easily do:

```sh
tee endpoints.json <<JSON
{
    "cloudformation": "http://localhost:4581",
    "s3": "http://localhost:4572"
}
JSON
cdk deploy --endpoints $(cat endpoints.json)
```

smilykoch commented:
@nmussy 
I totally agree with you, if that is the general pattern of the CDK.
Will change it right away.

Is there some helper parser, to support both comma separated and json input at the same time?

nmussy commented:
`--endpoints` would be the first object option for the CDK CLI as far as I can tell. I don't feel particularly strongly about having a shorthand syntax for it, maybe @RomainMuller can weigh it on it.

As far as implementation, you can always test if it's JSON object by try/catch-ing the parse, followed by a shorthand parse if it fails. Here's also the shorthand parser from the SDK CLI, in Python: [shorthand.py](https://github.com/aws/aws-cli/blob/develop/awscli/shorthand.py)

It's pretty hands on, I was just hoping to do the following, which wouldn't account for quotes:

```ts
Object.fromEntries('a=1,b=afijfif'.split(/,/g).map((entry) => entry.split('=')))
```

smilykoch commented:
@nmussy Than should do it.

`Object.fromEntries` is still not a part of typescript core, so i did it with a reduce

nmussy commented on packages/aws-cdk/lib/api/util/sdk.ts:
```diff
@@ -126,7 +124,19 @@ export class SDK implements ISDK {
+      try {
+        this.endpoints = JSON.parse(options.endpoints);
+      } catch (_e) {
+        this.endpoints = options.endpoints.split(/,/g).reduce((acc, entry) => {
```
You can use `.reduce<ServiceEndpoints>((acc, entry) =>` instead of 
`{} as ServiceEndpoints`

nmussy commented:
It's not TypeScript fault's, the CDK isn't targeting ES2019 (although I could have sworn `fromEntries` was older than that)

smilykoch commented:
Ohh, alright. Same result though :)

nmussy commented on packages/aws-cdk/lib/api/util/sdk.ts:
```diff
@@ -113,6 +146,7 @@ export class SDK implements ISDK {
     const environment = await this.resolveEnvironment(account, region);
     return new AWS.CloudFormation({
       ...this.retryOptions,
+      endpoint: this.endpoints.cloudformation,
```
You might get a `Cannot get property "cloudformation" of null` error with `--endpoints null`

smilykoch replied:
Should be fixed now.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* Result: SUCCEEDED
* [Build Logs](https://dfq49x5zyi.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=c3c760a7-0ffe-4315-8797-5c1a6c219b6f%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* Result: SUCCEEDED
* [Build Logs](https://dfq49x5zyi.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=3361a34e-8fe4-43b4-b2fe-3ec6cc075805%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* Result: SUCCEEDED
* [Build Logs](https://dfq49x5zyi.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=b762e628-1e76-4a80-999b-1cfad1e8cd08%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

RomainMuller commented on packages/aws-cdk/lib/api/util/sdk.ts:
```diff
@@ -53,6 +57,20 @@ export interface SDKOptions {
+   * cdk deploy --endpoints 'cloudformation=http://localhost:4581,s3=http://localhost:4572'
+   * cdk deploy --endpoints '{"cloudformation":"http://localhost:4581","s3":"http://localhost:4572"}'
+   *
+   * @default Online AWS cloud endpoints
```
I would call those *standard*, not *online*.

Also, it could be nice to allow users to specify these from a file, for example using a value prefixed by an `@`@. as is done by certain other tools (`--endpoints @./local-endpoints.json`).

smilykoch replied:
Hmm..

I must admit that i think its a lot of extra work for something that can be done through `--endpoints $(cat  ./local-endpoints.json)`, as your colleague suggested.

Regarding standard, i totally agree with you.

RomainMuller commented on packages/aws-cdk/lib/api/util/sdk.ts:
```diff
@@ -105,6 +124,20 @@ export class SDK implements ISDK {
+    if (options.endpoints) {
+      try {
+        this.endpoints = JSON.parse(options.endpoints) || {};
+      } catch (_e) {
```
I don't love exception-based control flow. Should be more intentional about what is valid - I'm afraid the fallback here is a little sloppy (it'll silently ignore a number of invalid cases, where it should have thrown instead).

smilykoch replied:
Hmm.. Would something like:

```
try {
    this.endpoints = JSON.parse(options.endpoints) || {};
} catch (_e) {}

this.endpoints = this.endpoints || options.endpoints.split(/,/g).reduce<ServiceEndpoints>((acc, entry) => {
    const [service, url] = entry.split('=');
    if (service && url) {
        acc[service] = url;
    }
    return acc;
}, {});
```

be better?
After all it would make it easier to add multiple try/catch cases in the future for other formats.

smilykoch commented:
I would definitely add the how to localstack, although there is still a bit of work in the localstack framework, as most of it is tested/implemented through new stacks instead of changesets.. So my next step is to make that work, and apply PR's to localstack and their dependencies. 

First step is this PR though.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* Result: FAILED
* [Build Logs](https://dfq49x5zyi.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=d3a1fd6d-6a7c-4343-b2f9-c3ebc52998e2%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* Result: FAILED
* [Build Logs](https://dfq49x5zyi.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=02f2a273-68b5-42b2-81ce-77f1d31c69fd%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* Result: FAILED
* [Build Logs](https://dfq49x5zyi.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=273fb920-2c99-4037-964d-272969013a72%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

amongil commented:
I'm trying to use the CDK with localstack. What is the status of this PR? Any help needed?

MathiasKoch commented:
@amongil This PR adds all the support needed in regards to CDK.

But there is still some way to go on the localstack end of things.. The main problem is that a lot of places in localstack, changesets through cloudformation seems to have been overlooked..

This means that there is a bit of work to be done in implementing the cloudformation mocks in the underlying MOTO python library. I began the work, but had to spend some time elsewhere. The main problem as i see it is the way localstack patches MOTO at runtime, making it a bit hard to follow what happens.. I guess if your used to python it helps a bit ;)

shivlaks commented:
do you need any help with this PR @MathiasKoch 
feel free to @ me when you're ready!

MathiasKoch commented:
@shivlaks Apart from the tests you just requested, i actually considered this PR as done.. 
I am not quite sure how i would properly test this feature though?

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* Result: FAILED
* [Build Logs](https://dfq49x5zyi.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=0d8d5e9d-181d-4c15-befc-e513c8116fb7%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

shivlaks commented:
> @shivlaks Apart from the tests you just requested, i actually considered this PR as done..
> I am not quite sure how i would properly test this feature though?

great. is there an integration test we could write?
Also, have you executed all the current integration tests and verified that they all pass?

There are also a couple of linting errors to take care of as the build is failing

shivlaks commented on packages/aws-cdk/lib/api/util/sdk.ts:
```diff
@@ -105,6 +124,25 @@ export class SDK implements ISDK {
+    if (options.endpoints) {
+      try {
+        this.endpoints = JSON.parse(options.endpoints);
+      } catch (_e) {}
```
please align this with the style of the code in this file i.e. `catch (e)`

why swallow up the exception? should we be logging something here?

MathiasKoch replied:
Well, the whole idea is that it should just try, and if it fails it should swallow it and try the next one, with a final fallback at the end. 
I don't think it makes sense to log anything. Also i can align with the style, but that would make a warning of unused variable.

MathiasKoch replied:
I can add a debug log in the catch (e)?

shivlaks commented on packages/aws-cdk/lib/api/util/sdk.ts:
```diff
@@ -105,6 +124,25 @@ export class SDK implements ISDK {
+              }
+              return acc;
+          }, {});
+      } catch (_e) {}
```
align with style of exceptions in reset of this file.

Why swallow up the exception? Should we be logging something here?

shivlaks commented on packages/aws-cdk/lib/api/util/sdk.ts:
```diff
@@ -53,6 +57,20 @@ export interface SDKOptions {
+   * cdk deploy --endpoints 'cloudformation=http://localhost:4581,s3=http://localhost:4572'
+   * cdk deploy --endpoints '{"cloudformation":"http://localhost:4581","s3":"http://localhost:4572"}'
+   *
+   * @default Standard AWS cloud endpoints
```
* what's does standard map to? there are multiple partitions to consider. a `@link` to service endpoints would be helpful as well

MathiasKoch commented:
@shivlaks 
I addressed your comments, and verified that all unit tests, and integration tests pass.

I would like to add a test for this feature, but i am not quite sure how.. I don't think an integration test makes sense, as the whole point of this feature is to NOT target the AWS cloud.

A unit test might make sense using mocks, but the mockSDK class in utils, seems to be very simple, and not include stuff like endpoints?

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* Result: SUCCEEDED
* [Build Logs](https://dfq49x5zyi.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=bbb4c712-06ff-43ce-8eab-5147fdb7a2c0%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

shivlaks commented:
> ### AWS CodeBuild CI Report
> * Result: SUCCEEDED
> * [Build Logs](https://dfq49x5zyi.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=bbb4c712-06ff-43ce-8eab-5147fdb7a2c0%2Fbuild.log) (available for 30 days)
> 
> _Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)_

good point, yes I think adding the unit tests that are relevant makes sense. You also have an exception so control flow could be tested there as well. Although you might not be able to verify the usage of the parameter in the SDK, you should still verify that the CLI command is able to parse and read input parameters as expected

dbartholomae commented:
Seeing that this is currently stuck, is there anything I can help to bring this to its close? Having localstack support in the CDK would be awesome :)

MathiasKoch commented:
@dbartholomae 
I think all it needs is unit tests, but unfortunately i don't have the time right now. You are more than welcome to take a swing at it.

micheleangioni commented:
Such a pity this is stuck at the moment. This would allow integration with LocalStack (and other libs) and would really help local development

mikelane commented:
+1 on getting this through the pipeline. I'll see if my team has any interest in adding unit tests.

ahammond commented:
We would find this super useful. I'll talk to our AWS account rep.

alexpulver commented:
This capability would also be helpful to use with AWS-released local development tools such as [Amazon DynamoDB Loca](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/DynamoDBLocal.html)l, [AWS Step Functions Local](https://docs.aws.amazon.com/step-functions/latest/dg/sfn-local.html), [AWS SAM CLI local commands](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/serverless-sam-cli-command-reference.html).

shivlaks commented:
I'll pick this issue up this week. Let's try and get this thing out soon.

shivlaks commented:
I don't have permissions to push to this branch so I'll start a new draft PR to port these changes over. We also refactored not too long ago so there's some minor shuffling to do.

shivlaks commented:
started a draft #7337 
I'll continue to finish off the remaining items in that one

MathiasKoch commented:
@shivlaks Thank you! Sorry i didn't have the time to finish this myself.

shivlaks commented:
> @shivlaks Thank you! Sorry i didn't have the time to finish this myself.

@MathiasKoch no worries! you did all the important groundwork here 🥳

shivlaks commented:
sorry for the mass mentions in advance: @alexpulver @ahammond @mikelane @micheleangioni @dbartholomae  @MathiasKoch 

Is `localstack` the primary use case or are there any other use cases. 

We're considering supporting a single endpoint since `localstack` supports it as of [`0.11.0`](https://github.com/localstack/localstack#overview)

That'd simplify this implementation, but if we need service specific endpoint configuration for any reason, please drop a note on #7337

mikelane commented:
> sorry for the mass mentions in advance: @alexpulver @ahammond @mikelane @micheleangioni @dbartholomae @MathiasKoch
> 
> Is `localstack` the primary use case or are there any other use cases.
> 
> We're considering supporting a single endpoint since `localstack` supports it as of [`0.11.0`](https://github.com/localstack/localstack#overview)
> 
> That'd simplify this implementation, but if we need service specific endpoint configuration for any reason, please drop a note on #7337

This is good stuff. We're trying to figure out how to test our stack locally. Since localstack doesn't work because of this issue, we have been trying to use Amplify. But darn it all, Amplify only sets up environments on a single hard-coded port, and of course, we have several services we're trying to set up and test.

Anyhow, given localstack 0.11.0's single endpoint hotness, we think that #7337 would meet our requirements.

ahammond commented:
We have been testing in a test sub-account. We are excited about this in the hopes of supporting testing CDK on dev laptops with localstack.

shivlaks commented:
awesome, I'm going to go with the single endpoint as @rix0rrr suggested (in #7337) - which is specific to localstack.

It'll be a bit simpler and i'll run through the end to end and update that PR shortly

MathiasKoch commented:
Closing in favour of https://github.com/aws/aws-cdk/pull/7337
